src/math.h

```diff
@@ -0,0 +1,45 @@
+// sgdnet: Penalized Generalized Linear Models with Stochastic Gradient Descent
+// Copyright (C) 2018  Johan Larsson
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#ifndef SGDNET_MATH_
+#define SGDNET_MATH_
+
+#include <RcppArmadillo.h>
+
+//' LogSumExp function
+//'
+//' @param x a vector in the log domain
+//'
+//' @return `log(sum(exp(x)))` while avoiding over/underflow.
+inline double LogSumExp(const arma::vec& x) {
+  double x_max = x.max();
+  return std::log(arma::accu(arma::exp(x - x_max))) + x_max;
+}
+
+//' Soft thresholding operator
+//'
+//' Solves \f$ \argmin_{x} 0.5||x - y||^{2} + \alpha ||x||_{1} \f$.
+//'
+//' @param x a double
+//' @param y a double
+//'
+//' @return
+inline double SoftThresholding(const double x,
+                               const double shrinkage) {
+  return std::max(x - shrinkage, 0.0) - std::max(-x - shrinkage, 0.0);
+}
```

Comment:
Should we make a "prox" object from which this inherits? That will make it easier to add other proximal operators (SCAD/MCP/group lasso) eventually

Reply:
Good idea. I have been thinking of something along those lines already.